curve-amm/src/lib.rs

```diff
@@ -0,0 +1,766 @@
+                .checked_add(&amounts[i])
+                .ok_or(CurveAmmError::Math)?;
+        }
+        let d1 = get_d(&new_balances, ann).ok_or(CurveAmmError::Math)?;
+        if d1 <= d0 {
+            return Err(CurveAmmError::WrongAssetAmount);
+        }
+        let mint_amount;
+        let mut fees = vec![FixedU128::zero(); n_coins];
+        // Only account for fees if we are not the first to deposit
+        if token_supply > zero {
+            // Deposit x + withdraw y would chargVe about same
+            // fees as a swap. Otherwise, one could exchange w/o paying fees.
+            // And this formula leads to exactly that equality
+            // fee = pool.fee * n_coins / (4 * (n_coins - 1))
+            let one = FixedU128::saturating_from_integer(1u8);
+            let four = FixedU128::saturating_from_integer(4u8);
+            let n_coins_f = Self::u128_to_fixed(n_coins as u128)?;
+            let fee_f: FixedU128 = pool.fee.into();
+            let n_coins_1 = n_coins_f.checked_sub(&one).ok_or(CurveAmmError::Math)?;
+            let four_n_coins_1 = four.checked_mul(&n_coins_1).ok_or(CurveAmmError::Math)?;
+            let pool_fees_n_coins = fee_f.checked_mul(&n_coins_f).ok_or(CurveAmmError::Math)?;
+            let fee_f = pool_fees_n_coins
+                .checked_div(&four_n_coins_1)
+                .ok_or(CurveAmmError::Math)?;
+            // let admin_fee_f: FixedU128 = pool.admin_fee.into();
+            for i in 0..n_coins {
+                let ideal_balance = d1
+                    .checked_mul(&old_balances[i])
+                    .and_then(|v| v.checked_div(&d0))
+                    .ok_or(CurveAmmError::Math)?;
+
+                let new_balance = new_balances[i];
+                // difference = abs(ideal_balance - new_balance)
+                let difference = if ideal_balance > new_balance {
+                    ideal_balance
+                        .checked_sub(&new_balance)
+                        .ok_or(CurveAmmError::Math)?
+                } else {
+                    new_balance
+                        .checked_sub(&ideal_balance)
+                        .ok_or(CurveAmmError::Math)?
+                };
+                fees[i] = fee_f.checked_mul(&difference).ok_or(CurveAmmError::Math)?;
+                new_balances[i] = new_balances[i]
+                    .checked_sub(&fees[i])
+                    .ok_or(CurveAmmError::Math)?;
+            }
+            let d2 = get_d(&new_balances, ann).ok_or(CurveAmmError::Math)?;
+
+            mint_amount = (|| {
+                token_supply
+                    .checked_mul(&d2.checked_sub(&d0)?)?
+                    .checked_div(&d0)
+            })()
+            .ok_or(CurveAmmError::Math)?;
+        } else {
+            mint_amount = d1;
+        }
+        if mint_amount < min_mint_amount {
+            return Err(CurveAmmError::RequiredAmountNotReached);
+        }
+
+        {
+            MUTEX.lock().await;
```

Comment:
MUTEX.lock() without variable drops immediately.

```suggestion
            let _lock = MUTEX.lock().await;
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a073316b4554db96da3cd12525dd9a66